File: src/Solutions/Euler017.hs
module Solutions.Euler017 where

first :: [String]
first = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
        "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen",
        "seventeen", "eighteen", "nineteen"]

dozens :: [String]
dozens = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]

itos :: Int -> String
itos n
    | n < 20 = first !! (n - 1)
    | n >= 100 = let prefix = first !! (n `div` 100 - 1) ++ "hundred" in
        if n `mod` 100 == 0 then prefix else prefix ++ "and" ++ itos (n `mod` 100)
    | otherwise = dozens !! (n `div` 10 - 2) ++ suffix where
        suffix = let d = n `mod` 10 in if d == 0 then "" else first !! (d - 1)

euler017 :: IO String
euler017 = return $ show $ length "onethousand" + sum (map (length . itos) [1..999])

File: rust/src/solutions/euler017.rs
fn itos(number: usize) -> String {
    let first = [
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen",
    ];
    let dozens = [
        "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety",
    ];
    match number {
        1000 => String::from("onethousand"),
        n if n < 20 => String::from(first[n - 1]),
        n if n >= 100 => {
            let prefix = String::from(first[n / 100 - 1]) + "hundred";
            let rem = n % 100;
            if rem == 0 {
                prefix
            } else {
                prefix + "and" + &itos(rem)
            }
        }
        _ => {
            let prefix = String::from(dozens[number / 10 - 2]);
            let rem = number % 10;
            if rem == 0 {
                prefix
            } else {
                prefix + first[rem - 1]
            }
        }
    }
}

pub fn euler017() -> String {
    (1..=1000).map(|n| itos(n).len()).sum::<usize>().to_string()
}
